File: cs/ex4_2_2.cs
using System.Collections.Generic;

abstract class Comparable<T> {
    public abstract bool gt(T that);
}

class Person : Comparable<Person> {
    int age;
    public override bool gt(Person that) { return this.age > that.age; }
}

class Example {
    static void sort<T>(List<T> lst) where T : Comparable<T> {
        if (lst[0].gt(lst[1])) {}
    }

    public static void Main() {
        List<Person> people = new List<Person>();
        sort<Person>(people);
    }
}

File: typescript/ex4_2_2.ts
abstract class Comparable<T> {
    abstract gt(that: T): boolean;
}

function sort<T extends Comparable<T>>(lst: Array<T>): void {
    if (lst[0].gt(lst[1])) {}
}

class Person extends Comparable<Person> {
    age: number = 0;
    gt(that: Person): boolean { return this.age > that.age; }
}

let people: Array<Person> = [];
sort<Person>(people);
